### ISSUE TITLE
immutable lists/maps with faster than linear time append/mutation

### ISSUE THREAD

**Issue by jmesserly** [google/built_collection.dart#44]
Immutable data structure suggestion:

In general, for an immutable tree, one can perform a "mutation" to any node in the graph in O(depth). Assuming the tree is roughly balanced, this ends up being O(lg(N)) were N is the total number of items in the tree. The "mutation" of course, is only seen in the newly returned tree. Original tree is unchanged. But they share as much structure as possible, to save time and space. This is the typical strategy for all languages that have immutable data structures.

For example, in Clojure, all of their immutable data structures took great pains to get efficient performance on changes. The array-like type used something like a B-tree with 32 as the branching factor. So for example, you can "mutate" a 1000 element array and it only rewrites 2 little Java arrays. (Disclaimer: it was like 5 yrs ago when I last looked, details may have changed.)

Similarly it would be possible to have BuiltMap that supports "mutation", returning a new BuiltMap that shares almost all of the structure with the original map.

Anyway, the reason I bring this up, I've seen a lot of folks use BuiltList/BuiltMap like:

``` dart
void addItem(Item item) {
  _myBuiltList = _myBuiltList.rebuild((b) => b.add(item));
}
```

(edit: fixed example to make more sense)

That can easily lead to quadratic code. If you call addItem N times, your calling code will be O(N^2). At least this is what I gathered from reading the current implementation code for BuiltList/Map--that rebuild is O(N).

Anyway if this sounds like I good idea, I might try and prototype something, but I wanted to file this first to check my understanding. (p.s. <3 immutable data structures. this package is super useful.)

**Comment by davidmorgan** [comment]
Thanks John. Indeed, the idea was to keep the possibility of using such data structures open.

One reason we haven't looked at it yet is we don't really have a good use case to benchmark with.

The "addItem" example might look like a problem, but so far it's always been in UI code where adding an item corresponds to a click and subsequent redraw. In that case the rendering time is likely to be far higher than the list copy, so it's not a big concern. For larger updates that are caused by a single click, you're supposed to use builders rather than repeatedly rebuilding.

One other direction that might be worth exploring is based on the idea that in the most common use of 'rebuild', the original list may not ever be used again. So for "addItem" we could do this instead:

1) update the old BuiltList to say "I'm the new BuiltList minus the last item"; this is expensive if it's used, but we assume it probably won't be
2) create the new BuiltList by stealing the underlying list from the first and adding the item

Then, if the old list is never used again, we do nothing extra, and it's almost as fast as a mutable list to update, and as fast as a mutable list for reading. If the old list _is_ used again then extra work is needed to either implement the list interface over a "BuiltList minus last item", or we just copy the underlying list and actually remove the last item.

In order to pursue this line of thought we'd also need a good use case to benchmark with.

So, bottom line is, let's find some code that actually wants fast immutable data structures :) and then, yes, please do go ahead and prototype something, I'd love to see it.

**Comment by pinyin** [comment]
What about use cases like [flutter_redux](https://pub.dartlang.org/packages/flutter_redux)? Things like HAMT would be quite helpful when one want to update/diff a large amount of data between Flutter frames.

**Comment by davidmorgan** [comment]
Maybe--I'd love to see benchmarks.

Note that `built_value` is already optimized so that updating a deeply nested structure of objects is O(length of path from modification to root), and not O(size of data structure).
